File: apps/api/app/db/models/workspace_layouts.py
from __future__ import annotations

from datetime import UTC, datetime
from uuid import uuid4

from sqlalchemy import Boolean, DateTime, JSON, String
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class WorkspaceLayoutRecord(Base):
    __tablename__ = "workspace_layouts"

    layout_id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: f"wsl-{uuid4().hex[:16]}")
    owner_user_id: Mapped[str] = mapped_column(String(255), default="default", index=True)
    scope_type: Mapped[str] = mapped_column(String(128), index=True)
    scope_id: Mapped[str] = mapped_column(String(255), index=True)
    layout_kind: Mapped[str] = mapped_column(String(128), index=True)
    layout_role: Mapped[str] = mapped_column(String(64), default="named_snapshot", index=True)
    name: Mapped[str] = mapped_column(String(255), default="未命名布局")
    is_default: Mapped[bool] = mapped_column(Boolean, default=False)
    payload_schema_version: Mapped[str] = mapped_column(String(128))
    payload: Mapped[dict] = mapped_column(JSON, default=dict)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(UTC))
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(UTC))
    last_used_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(UTC))

File: apps/web/src/lib/workspaceLayouts.ts
import { api } from "./api";

export type WorkspaceLayoutRecord = {
  layout_id: string;
  owner_user_id: string;
  scope_type: string;
  scope_id: string;
  layout_kind: string;
  layout_role: string;
  name: string;
  is_default: boolean;
  payload_schema_version: string;
  payload: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  last_used_at: string;
};

export type WorkspaceLayoutEnvelope = {
  items: WorkspaceLayoutRecord[];
};

export type WorkspaceLayoutListParams = {
  owner_user_id?: string;
  scope_type: string;
  scope_id: string;
  layout_kind: string;
  layout_role?: string;
};

export type WorkspaceLayoutCreateInput = {
  owner_user_id?: string;
  scope_type: string;
  scope_id: string;
  layout_kind: string;
  layout_role: string;
  name: string;
  is_default?: boolean;
  payload_schema_version: string;
  payload: Record<string, unknown>;
};

export type WorkspaceLayoutCurrentInput = {
  owner_user_id?: string;
  scope_type: string;
  scope_id: string;
  layout_kind: string;
  name: string;
  payload_schema_version: string;
  payload: Record<string, unknown>;
};

export async function listWorkspaceLayouts(params: WorkspaceLayoutListParams): Promise<WorkspaceLayoutEnvelope> {
  const response = await api.get<WorkspaceLayoutEnvelope>("/workspace-layouts", {
    params: {
      owner_user_id: params.owner_user_id ?? "default",
      scope_type: params.scope_type,
      scope_id: params.scope_id,
      layout_kind: params.layout_kind,
      ...(params.layout_role ? { layout_role: params.layout_role } : {}),
    },
  });
  return response.data;
}

export async function createWorkspaceLayout(payload: WorkspaceLayoutCreateInput): Promise<WorkspaceLayoutRecord> {
  const response = await api.post<WorkspaceLayoutRecord>("/workspace-layouts", {
    ...payload,
    owner_user_id: payload.owner_user_id ?? "default",
    is_default: payload.is_default ?? false,
  });
  return response.data;
}

export async function upsertCurrentWorkspaceLayout(payload: WorkspaceLayoutCurrentInput): Promise<WorkspaceLayoutRecord> {
  const response = await api.put<WorkspaceLayoutRecord>("/workspace-layouts/current", {
    ...payload,
    owner_user_id: payload.owner_user_id ?? "default",
  });
  return response.data;
}

export async function deleteWorkspaceLayout(layoutId: string): Promise<{ deleted_layout_id: string }> {
  const response = await api.delete<{ deleted_layout_id: string }>(`/workspace-layouts/${layoutId}`);
  return response.data;
}
